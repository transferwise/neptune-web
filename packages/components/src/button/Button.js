import React from 'react';
import Types from 'prop-types';
import classNames from 'classnames';
import requiredIf from 'react-required-if';
import { logActionRequiredIf } from '../utilities';

import './Button.css';

import { Size, Type } from '../common';

const Button = ({
  className,
  block,
  children,
  disabled,
  htmlType,
  loading,
  size,
  type,
  ...rest
}) => {
  const classes = classNames(
    `btn btn-${size}`,
    `np-btn np-btn-${size}`,
    {
      'btn-loading': loading,
      'btn-primary': type === Type.PRIMARY,
      'btn-success': type === Type.PAY,
      'btn-default': type === Type.SECONDARY,
      'btn-danger': type === Type.DANGER,
      'btn-link': type === Type.LINK,
      'btn-block np-btn-block': block,
    },
    className,
  );

  logActionRequiredIf(
    'Button has deprecated the `Button.Size.EXTRA_SMALL` value for the `size` prop. Please use Button.Size.SMALL instead.',
    size === Size.EXTRA_SMALL,
  );
  return (
    /* eslint-disable react/button-has-type */
    <button type={htmlType} className={classes} disabled={disabled || loading} {...rest}>
      {children}
      {loading && <span className={classNames('btn-loader', { 'm-l-2': !block })} />}
    </button>
  );
};

Button.Type = Type;
Button.Size = {
  EXTRA_SMALL: Size.EXTRA_SMALL,
  SMALL: Size.SMALL,
  MEDIUM: Size.MEDIUM,
  LARGE: Size.LARGE,
};

Button.propTypes = {
  className: Types.string,
  type: Types.oneOf([
    Button.Type.PRIMARY,
    Button.Type.PAY,
    Button.Type.SECONDARY,
    Button.Type.DANGER,
    Button.Type.LINK,
  ]),
  /** @DEPRECATED Button.Size.EXTRA_SMALL */
  size: Types.oneOf([
    Button.Size.EXTRA_SMALL,
    Button.Size.SMALL,
    Button.Size.MEDIUM,
    Button.Size.LARGE,
  ]),
  disabled: Types.bool,
  block: Types.bool,
  loading: Types.bool,
  // eslint-disable-next-line
  onClick: requiredIf(Types.func, (props) => props.htmlType !== 'submit'),
  children: Types.node.isRequired,
  htmlType: Types.oneOf(['submit', 'reset', 'button']),
};

Button.defaultProps = {
  className: null,
  size: Button.Size.MEDIUM,
  type: Button.Type.PRIMARY,
  disabled: false,
  block: false,
  loading: false,
  htmlType: 'button',
};

export default Button;
