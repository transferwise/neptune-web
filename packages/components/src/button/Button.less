.np-btn {
  position: relative;

  &.np-btn-block > span.btn-loader {
    position: absolute;
    left: var(--size-24);
  }

  &.np-btn-xs > span.btn-loader {
    background-size: 16px 16px;
  }

  &.np-btn-xs.btn-block > span.btn-loader {
    top: 0;
  }
}
