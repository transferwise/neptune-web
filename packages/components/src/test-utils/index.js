import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
