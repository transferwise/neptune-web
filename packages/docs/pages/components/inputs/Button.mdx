import { LiveEditorBlock, GeneratePropsTable } from '../../../utils';
import { Button } from '@transferwise/components';
import code from '../../../liveEditorCode/button.code';

<LiveEditorBlock code={code} scope={{ Button }} />
<GeneratePropsTable componentName="Button" />

## Variants

<div className="scroll-table">
  <table className="docs-table table ">
    <thead>
      <tr>
        <th>Type</th>
        <th>Priority</th>
        <th>Preview</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Accent</th>
        <th>Primary</th>
        <td>
          <Button block={true} onClick={() => {}}>
            Continue
          </Button>
        </td>
      </tr>
      <tr>
        <th>Accent</th>
        <th>Secondary</th>
        <td>
          <Button priority="secondary" block={true} onClick={() => {}}>
            Convert money
          </Button>
        </td>
      </tr>
      <tr>
        <th>Accent</th>
        <th>Tertiary</th>
        <td>
          <Button priority="tertiary" block={true} onClick={() => {}}>
            Convert money
          </Button>
        </td>
      </tr>
      <tr>
        <th>Positive</th>
        <th>Primary</th>
        <td>
          <Button type="positive" block={true} onClick={() => {}}>
            Send money
          </Button>
        </td>
      </tr>
      <tr>
        <th>Positive</th>
        <th>Secondary</th>
        <td>
          <Button type="positive" priority="secondary" block={true} onClick={() => {}}>
            Confirm
          </Button>
        </td>
      </tr>
      <tr>
        <th>Negative</th>
        <th>Primary</th>
        <td>
          <Button type="negative" block={true} onClick={() => {}}>
            Delete
          </Button>
        </td>
      </tr> <tr>
        <th>Negative</th>
        <th>Secondary</th>
        <td>
          <Button type="negative" priority="secondary" block={true} onClick={() => {}}>
            Cancel
          </Button>
        </td>
      </tr>
    </tbody>
  </table>
</div>

export const meta = {
  name: 'Button',
};
