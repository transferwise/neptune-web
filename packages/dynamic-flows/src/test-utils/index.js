import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';
import { act } from 'react-dom/test-utils';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount, getMockFetchPromise, wait };
