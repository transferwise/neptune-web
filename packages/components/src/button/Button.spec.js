/* eslint-disable no-console */
import React from 'react';
import '@testing-library/jest-dom';
import { render, cleanup, screen, userEvent } from '../test-utils';

import Button from '.';

const { ACCENT, POSITIVE, NEGATIVE } = Button.Type;
const { PRIMARY, SECONDARY, TERTIARY } = Button.Priority;
const { SMALL, MEDIUM, LARGE } = Button.Size;

describe('Button', () => {
  const props = {
    onClick: jest.fn(),
    children: 'Send money',
  };

  const origWarn = console.warn;
  let mockedWarn;

  beforeAll(() => {
    mockedWarn = jest.fn();
    console.warn = mockedWarn;
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  afterAll(() => {
    console.warn = origWarn;
  });

  describe('by default', () => {
    beforeEach(() => {
      render(<Button {...props} />);
    });

    it('renders the text', () => {
      screen.getByText('Send money');
    });

    it('is not disabled', () => {
      expect(screen.getByRole('button')).not.toBeDisabled();
    });

    it('renders a medium button of type accent and priority primary', () => {
      expect(render(<Button {...props} />).container).toMatchSnapshot();
    });
  });

  describe('button attributes', () => {
    it('disables the button if disabled', () => {
      render(<Button {...props} disabled />);
      expect(screen.getByRole('button')).toBeDisabled();
    });

    it('sets the htmlType if set', () => {
      render(<Button {...props} htmlType="submit" />);
      expect(screen.getByRole('button')).toHaveAttribute('type', 'submit');
    });

    it('passes through custom classes if set', () => {
      render(<Button {...props} className="catsarethebest" />);
      expect(screen.getByRole('button')).toHaveClass('catsarethebest');
    });
  });

  describe('onClick', () => {
    it('calls onClick when clicked', () => {
      const onClick = jest.fn();
      render(<Button {...props} onClick={onClick} />);
      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('does not call onClick when clicked if disabled', () => {
      const onClick = jest.fn();
      render(<Button {...props} onClick={onClick} disabled />);
      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(0);
    });
  });

  describe('sizes', () => {
    it('renders small buttons', () => {
      expect(render(<Button {...props} size={SMALL} />).container).toMatchSnapshot();
    });

    it('renders medium buttons', () => {
      expect(render(<Button {...props} size={MEDIUM} />).container).toMatchSnapshot();
    });

    it('renders large buttons', () => {
      expect(render(<Button {...props} size={LARGE} />).container).toMatchSnapshot();
    });
  });

  describe('types', () => {
    it('renders accent buttons', () => {
      expect(render(<Button {...props} type={ACCENT} />).container).toMatchSnapshot();
    });

    it('renders positive buttons', () => {
      expect(render(<Button {...props} type={POSITIVE} />).container).toMatchSnapshot();
    });

    it('renders negative buttons', () => {
      expect(render(<Button {...props} type={NEGATIVE} />).container).toMatchSnapshot();
    });
  });

  describe('priorities', () => {
    it('renders primary buttons', () => {
      [ACCENT, POSITIVE, NEGATIVE].forEach((type) =>
        expect(
          render(<Button {...props} priority={PRIMARY} type={type} />).container,
        ).toMatchSnapshot(),
      );
    });

    it('renders secondary buttons', () => {
      [ACCENT, POSITIVE, NEGATIVE].forEach((type) =>
        expect(
          render(<Button {...props} priority={SECONDARY} type={type} />).container,
        ).toMatchSnapshot(),
      );
    });

    it('renders tertiary buttons', () => {
      expect(
        render(<Button {...props} priority={TERTIARY} type={ACCENT} />).container,
      ).toMatchSnapshot();
    });

    it('defaults tertiary buttons to secondary for positive buttons', () => {
      [POSITIVE, NEGATIVE].forEach((type) =>
        expect(
          render(<Button {...props} priority={TERTIARY} type={type} />).container,
        ).toMatchSnapshot(),
      );
    });
  });

  describe('alternative states', () => {
    it('renders as loading if loading is true', () => {
      const { container } = render(<Button {...props} loading />);
      expect(container).toMatchSnapshot();

      expect(container.querySelector('.btn-loader')).toBeInTheDocument();
    });

    it('renders as block if block is true', () => {
      expect(render(<Button {...props} block />).container).toMatchSnapshot();
    });
  });

  describe('deprecated types', () => {
    it('renders primary as accent buttons and logs a warning ', () => {
      expect(render(<Button {...props} type={Button.Type.PRIMARY} />).container).toMatchSnapshot();
      expect(mockedWarn).toHaveBeenCalledTimes(1);
    });

    it('renders pay as positive buttons and logs a warning', () => {
      expect(render(<Button {...props} type={Button.Type.PAY} />).container).toMatchSnapshot();
      expect(mockedWarn).toHaveBeenCalledTimes(1);
    });

    it('renders danger as negative buttons with priority secondary and logs a warning', () => {
      expect(render(<Button {...props} type={Button.Type.DANGER} />).container).toMatchSnapshot();
      expect(mockedWarn).toHaveBeenCalledTimes(1);
    });

    it('renders link as accent buttons with priority tertiary and logs a warning', () => {
      expect(render(<Button {...props} type={Button.Type.LINK} />).container).toMatchSnapshot();
    });
  });
});
