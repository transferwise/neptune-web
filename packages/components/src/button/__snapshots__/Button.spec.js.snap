// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button alternative states renders as block if block is true 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-block np-btn-block btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button alternative states renders as loading if loading is true 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-loading btn-accent btn-priority-1"
    disabled=""
    type="button"
  >
    Send money
    <span
      class="btn-loader m-l-2"
    />
  </button>
</div>
`;

exports[`Button by default renders a medium button of type accent and priority primary 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button deprecated types renders danger as negative buttons with priority secondary and logs a warning 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-negative btn-priority-2"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button deprecated types renders link as accent buttons with priority tertiary and logs a warning 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-3"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button deprecated types renders pay as positive buttons and logs a warning 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-positive btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button deprecated types renders primary as accent buttons and logs a warning  1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities defaults tertiary buttons to secondary for positive buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-positive btn-priority-2"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities defaults tertiary buttons to secondary for positive buttons 2`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-negative btn-priority-2"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities renders primary buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities renders primary buttons 2`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-positive btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities renders primary buttons 3`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-negative btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities renders secondary buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-2"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities renders secondary buttons 2`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-positive btn-priority-2"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities renders secondary buttons 3`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-negative btn-priority-2"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button priorities renders tertiary buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-3"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button sizes renders large buttons 1`] = `
<div>
  <button
    class="btn btn-lg np-btn np-btn-lg btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button sizes renders medium buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button sizes renders small buttons 1`] = `
<div>
  <button
    class="btn btn-sm np-btn np-btn-sm btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button types renders accent buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-accent btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button types renders negative buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-negative btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;

exports[`Button types renders positive buttons 1`] = `
<div>
  <button
    class="btn btn-md np-btn np-btn-md btn-positive btn-priority-1"
    type="button"
  >
    Send money
  </button>
</div>
`;
